clickhouse/columns/map.h

```diff
@@ -187,13 +191,16 @@ class ColumnMapT : public ColumnMap {
             if (size() != other.size()) {
                 return false;
             }
-            using Vector = std::vector<std::pair<Key, Value>>;
-            Vector l(begin(), end());
-            Vector r(other.begin(), other.end());
-            auto comp = [](const auto& l, const auto& r) { return l.frist < r.first; };
-            std::sort(l.begin(), l.end(), comp);
-            std::sort(r.begin(), r.end(), comp);
-            return std::equal(l.begin(), l.end(), r.begin(), r.end());
+            const auto make_index = [](const auto& data) {
+                std::vector<size_t> result{data.Size()};
+                std::generate(result.begin(), result.end(), [i = 0] () mutable { return i++; });
+                std::sort(result.begin(), result.end(), [&data](size_t l, size_t r) {return data[l] < data[r];});
+                return result;
+            };
```

Comment:
Not entirely clear peace of code, so could you please add some comments to it?

Another thing: this allocates `(2 * N)*sizeof(size_t)` extra memory on the heap. IMO it could be more beneficial to create a sorted array (of pointers/references) from left items and then do a binary-search lookups with items from the right.

That way the algorithmic complexity stays the same `O(N*logN)`, but memory requirement drops significantly: only `N*sizeof(size_t)` instead of `2N*sizeof(t)`.

Anyway, it would be really nice to see some numbers on that, could you please add an extra entry to `clickhouse-cpp/ut/performance_tests.cpp` ?